File: Back-End/backend/src/models/PlanConfig.ts
import * as admin from 'firebase-admin';
import { db } from '../global/firebase';

// ── 型別定義 ────────────────────────────────────────────────────────────────

export interface PlanConfigInput {
  annualInvest:        number;                  // 每年投入預設值
  rBase:               number;                  // 基礎殖利率 raw decimal（0.08 = 8%）
  inflation:           'low' | 'base' | 'high'; // 通膨情境
  kRisk:               number;                  // 景氣係數
  startYear:           number;                  // 起始年份
  overrides:           Record<string, number>;  // { "1": 150000 }，yearIndex（1-based）→ 該年計畫投入
  currentYearReinvest: number;                  // 當年度手動填入的再投入金額
}

export interface PlanConfigDoc extends PlanConfigInput {
  updatedAt: string; // ISO string
}

const COL      = 'plan_config';
const DOC_ID   = 'main';

const DEFAULTS: PlanConfigInput = {
  annualInvest:        120000,
  rBase:               0.08,
  inflation:           'base',
  kRisk:               1.0,
  startYear:           new Date().getFullYear(),
  overrides:           {},
  currentYearReinvest: 0,
};

// ── Model ───────────────────────────────────────────────────────────────────

export class PlanConfig {
  /** 取得計畫設定；Firestore 無資料時回傳預設值 */
  static async find(): Promise<PlanConfigDoc> {
    const doc = await db.collection(COL).doc(DOC_ID).get();
    if (!doc.exists) {
      return { ...DEFAULTS, updatedAt: new Date().toISOString() };
    }
    return deserialize(doc);
  }

  /** 整筆覆寫計畫設定 */
  static async upsert(input: PlanConfigInput): Promise<PlanConfigDoc> {
    const ref = db.collection(COL).doc(DOC_ID);
    await ref.set({
      annual_invest:          input.annualInvest,
      r_base:                 input.rBase,
      inflation:              input.inflation,
      k_risk:                 input.kRisk,
      start_year:             input.startYear,
      overrides:              input.overrides,
      current_year_reinvest:  input.currentYearReinvest,
      updated_at:             admin.firestore.FieldValue.serverTimestamp(),
    });
    const updated = await ref.get();
    return deserialize(updated);
  }
}

// ── 反序列化 ─────────────────────────────────────────────────────────────────

function deserialize(doc: admin.firestore.DocumentSnapshot): PlanConfigDoc {
  const d = doc.data()!;
  const ua = d['updated_at'];
  return {
    annualInvest:        d['annual_invest']         ?? DEFAULTS.annualInvest,
    rBase:               d['r_base']                ?? DEFAULTS.rBase,
    inflation:           d['inflation']             ?? DEFAULTS.inflation,
    kRisk:               d['k_risk']               ?? DEFAULTS.kRisk,
    startYear:           d['start_year']            ?? DEFAULTS.startYear,
    overrides:           d['overrides']             ?? {},
    currentYearReinvest: d['current_year_reinvest'] ?? 0,
    updatedAt: ua instanceof admin.firestore.Timestamp
      ? ua.toDate().toISOString()
      : new Date().toISOString(),
  };
}

File: Back-End/shioaji-service/models/plan_config.py
from __future__ import annotations
import asyncio
from typing import Optional
from google.cloud.firestore import SERVER_TIMESTAMP
from lib.firebase import get_db

COL = "plan_config"
DOC_ID = "main"

_DEFAULTS = {
    "annualInvest": 120000,
    "rBase": 0.08,
    "inflation": "base",
    "kRisk": 1.0,
    "startYear": None,  # filled at runtime
    "overrides": {},
    "currentYearReinvest": 0,
}


def _ts_to_iso(ts) -> Optional[str]:
    return ts.isoformat() if hasattr(ts, "isoformat") else None


def _from_doc(doc) -> dict:
    from datetime import datetime
    d = doc.to_dict()
    return {
        "annualInvest": d.get("annual_invest", _DEFAULTS["annualInvest"]),
        "rBase": d.get("r_base", _DEFAULTS["rBase"]),
        "inflation": d.get("inflation", _DEFAULTS["inflation"]),
        "kRisk": d.get("k_risk", _DEFAULTS["kRisk"]),
        "startYear": d.get("start_year", datetime.now().year),
        "overrides": d.get("overrides", {}),
        "currentYearReinvest": d.get("current_year_reinvest", 0),
        "updatedAt": _ts_to_iso(d.get("updated_at")),
    }


async def find() -> dict:
    def _run():
        from datetime import datetime
        doc = get_db().collection(COL).document(DOC_ID).get()
        if not doc.exists:
            return {**_DEFAULTS, "startYear": datetime.now().year, "updatedAt": None}
        return _from_doc(doc)
    return await asyncio.to_thread(_run)


async def upsert(data: dict) -> dict:
    def _run():
        ref = get_db().collection(COL).document(DOC_ID)
        ref.set({
            "annual_invest": data["annualInvest"],
            "r_base": data["rBase"],
            "inflation": data["inflation"],
            "k_risk": data["kRisk"],
            "start_year": data["startYear"],
            "overrides": data.get("overrides", {}),
            "current_year_reinvest": data.get("currentYearReinvest", 0),
            "updated_at": SERVER_TIMESTAMP,
        })
        return _from_doc(ref.get())
    return await asyncio.to_thread(_run)
